vcd/resource_vcd_nsxt_route_advertisement_test.go

```diff
@@ -0,0 +1,368 @@
+//go:build network || nsxt || ALL || functional
+// +build network nsxt ALL functional
+
+package vcd
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"regexp"
+	"strconv"
+	"testing"
+)
+
+func TestAccVcdNsxtRouteAdvertisement(t *testing.T) {
+	preTestChecks(t)
+	skipNoNsxtConfiguration(t)
+
+	isRouteAdvertisementEnable := true
+	subnet1 := "192.168.1.0/24"
+	subnet2 := "192.168.2.0/24"
+
+	// String map to fill the template
+	var params = StringMap{
+		"Name":              t.Name(),
+		"Org":               testConfig.VCD.Org,
+		"NsxtVdc":           testConfig.Nsxt.Vdc,
+		"NsxtVdcGroup":      testConfig.Nsxt.VdcGroup,
+		"EdgeGw":            testConfig.Nsxt.EdgeGateway,
+		"EdgeGwVdcGroup":    testConfig.Nsxt.VdcGroupEdgeGateway,
+		"Enabled":           strconv.FormatBool(isRouteAdvertisementEnable),
+		"Subnet1Cidr":       subnet1,
+		"Subnet2Cidr":       subnet2,
+		"NsxtImportSegment": testConfig.Nsxt.NsxtImportSegment,
+	}
+
+	configText1 := templateFill(testAccNsxtRouteAdvertisementCreation, params)
+	debugPrintf("#[DEBUG] CONFIGURATION for step 1: %s", configText1)
+
+	params["FuncName"] = t.Name() + "-step2"
+	configText2 := templateFill(testAccNsxtRouteAdvertisementUpdate, params)
+	debugPrintf("#[DEBUG] CONFIGURATION for step 2: %s", configText2)
+
+	params["FuncName"] = t.Name() + "-step3"
+	configText3 := templateFill(testAccNsxtRouteAdvertisementDisabled, params)
+	debugPrintf("#[DEBUG] CONFIGURATION for step 3: %s", configText3)
+
+	if vcdShortTest {
+		t.Skip(acceptanceTestsSkipped)
+		return
+	}
+
+	// Ensure Edge Gateway has a dedicated Tier 0 gateway (External network) as BGP and Route
+	// Advertisement configuration requires it. Restore it right after the test so that other
+	// tests are not impacted.
+	updateEdgeGatewayTier0Dedication(t, true)
+	defer updateEdgeGatewayTier0Dedication(t, false)
+
+	resource.Test(t, resource.TestCase{
+		ProviderFactories: testAccProviders,
+		PreCheck:          func() { testAccPreCheck(t) },
+		CheckDestroy:      testAccCheckNsxtRouteAdvertisement(testConfig.Nsxt.Vdc, testConfig.Nsxt.EdgeGateway),
+		Steps: []resource.TestStep{
+			{
+				Config: configText1,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestMatchResourceAttr("vcd_nsxt_route_advertisement.testing", "id", regexp.MustCompile(`^urn:vcloud:gateway:.*$`)),
+					resource.TestCheckResourceAttr("vcd_nsxt_route_advertisement.testing", "enabled", strconv.FormatBool(isRouteAdvertisementEnable)),
+					resource.TestCheckResourceAttr("vcd_nsxt_route_advertisement.testing", "subnets.#", "1"),
+					resource.TestCheckResourceAttr("vcd_nsxt_route_advertisement.testing", "subnets.0", subnet1),
+				),
+			},
+			{
```

Comment:
Can you pull in additional test step which check that data source functions and returns the same fields which were set. You can look for inspiration how to do this in many places, but here is one I remember doing so:
https://github.com/vmware/terraform-provider-vcd/blob/main/vcd/resource_vcd_nsxt_alb_edgegateway_service_engine_group_test.go#L177

Reply:
All right, I've added a new test step for testing datasource. This new test step is the number 3. The former number 3 has been renamed to 4. 

The test step can be found here: https://github.com/mikeletux/terraform-provider-vcd/blob/c1690c2df11da7fa331913ce10e64e456c7a81e6/vcd/resource_vcd_nsxt_route_advertisement_test.go#L87-L97

The template being used comes with the `# skip-binary-test` directive to avoid including it in the binary test as suggested: https://github.com/mikeletux/terraform-provider-vcd/blob/c1690c2df11da7fa331913ce10e64e456c7a81e6/vcd/resource_vcd_nsxt_route_advertisement_test.go#L155-L160